x/incentives/keeper/hooks_test.go

```diff
@@ -352,11 +362,154 @@ func (s *KeeperTestSuite) Test_AfterEpochEnd_Group_ChangeVolumeBetween() {
+	poolIDToExpectedDistributionMap := s.computeExpectedDistributonAmountsFromVolume(defaultCoins, poolIDToVolumeMap, volumeA)
+
+	// System under test
+	err = s.App.IncentivesKeeper.AfterEpochEnd(s.Ctx, distrEpochIdentifier, 1)
+	s.Require().NoError(err)
+
+	s.validateDistributionForGroup(poolIDsGroup, poolIDToExpectedDistributionMap)
+
+	// Create perpetual group distributing to the same pool (for ease of setup)
+	_, err = s.App.IncentivesKeeper.CreateGroup(s.Ctx, defaultCoins, types.PerpetualNumEpochsPaidOver, s.TestAccs[0], poolIDsGroup)
+	s.Require().NoError(err)
+
+	s.increaseVolumeForPools(poolIDsGroup, equalPoolVolumes)
+
+	// Compute the expected
+	// Since we have a non-perpetual gauge with 2x defaultCoins and a perpetual gauge with 1x defaultCoins,
+	// The final total is 3x the expected for the non-perpetual gauge in the first epoch.
+	// Merge operations below essentially do the 3x multiplication.
+	expectedFinalCoinsDistributed := osmoutils.MergeCoinMaps(poolIDToExpectedDistributionMap, poolIDToExpectedDistributionMap)
+	expectedFinalCoinsDistributed = osmoutils.MergeCoinMaps(expectedFinalCoinsDistributed, poolIDToExpectedDistributionMap)
+
+	// System under test
+	err = s.App.IncentivesKeeper.AfterEpochEnd(s.Ctx, distrEpochIdentifier, 2)
+	s.Require().NoError(err)
+
+	// Group should distribute expected amounts.
+	s.validateDistributionForGroup(poolIDsGroup, expectedFinalCoinsDistributed)
+}
+
+// This test focuses on configuring volume by swapping instead of using
+// a direct volume setter helper in poolmanager contrary to all other tests.
+// Since we track volume in bond denom (OSMO), we first setup2 pools that are paired with the bond denom.
+// Next, we setup two pools that are to be packaged in group. One of the tokens in the pool is a token that is also
+// paired with bond denom in one of the first 2 pools.
+// Increase volume by swapping in the second pair of pools.
+// Create a group with the second pair of pools.
+// Increase volume by swapping in the second pair of pools.
+// Call AfterEpochEnd hook.
+// Validate that the distribution is correct.
+func (s *KeeperTestSuite) Test_AfterEpochEnd_Group_SwapAndDistribute() {
+	// Setup UOSMO as bond denom
+	stakingParams := s.App.StakingKeeper.GetParams(s.Ctx)
+	stakingParams.BondDenom = UOSMO
+	s.App.StakingKeeper.SetParams(s.Ctx, stakingParams)
+
+	// Create UOSMO / USDC pool
+	s.PrepareCustomBalancerPool([]balancer.PoolAsset{
+		{Token: sdk.NewCoin(UOSMO, defaultAmount), Weight: sdk.OneInt()},
+		{Token: sdk.NewCoin(USDC, defaultAmount), Weight: sdk.OneInt()},
+	}, balancer.PoolParams{
+		SwapFee: osmomath.ZeroDec(),
+		ExitFee: osmomath.ZeroDec(),
+	})
+
+	// Create UOSMO / BAR pool
+	s.PrepareCustomBalancerPool([]balancer.PoolAsset{
+		{Token: sdk.NewCoin(UOSMO, defaultAmount), Weight: sdk.OneInt()},
+		{Token: sdk.NewCoin(BAR, defaultAmount), Weight: sdk.OneInt()},
+	}, balancer.PoolParams{
+		SwapFee: osmomath.ZeroDec(),
+		ExitFee: osmomath.ZeroDec(),
+	})
```

Comment:
Not a requirement, but consider making the second pool a different pool type, to just expand the testing surface.

Reply:
There are other tests that already cover all pool types. I specifically isolate all variations into separate test fixture to keep the focus on the main goal of each test and have everything else constants / simplified